app/assets/javascripts/discourse/app/controllers/discovery/topics.js

```diff
@@ -72,23 +72,37 @@ const controllerOpts = {
       this.set("discovery.loading", true);
 
       this.topicTrackingState.resetTracking();
+      let self = this;
+
       this.store.findFiltered("topicList", { filter }).then((list) => {
         TopicList.hideUniformCategory(list, this.category);
 
-        this.setProperties({ model: list });
-        this.resetSelected();
-
-        if (this.topicTrackingState) {
-          this.topicTrackingState.sync(list, filter);
+        // If query params are present in the current route, we need still need to sync topic
+        // tracking with the topicList without any query params. Then we set the topic
+        // list to the list filtered with query params in the afterRefresh.
+        const params = this.router.currentRoute.queryParams;
+        if (Object.keys(params).length) {
```

Comment:
This seems a bit weird to me. You are querying the topic list, then after it's loaded you load it again if there are query parameters? Couldn't you load it once with the query parameters and avoid all this?

Reply:
This is the tricky part. Topic tracking state needs the topic list without query parameters. If query params are present, then fetch the topic list with query params for the list model. It's kinda hacky, but the topic tracking state needs to sync to the un-query-param'd topic list here https://github.com/discourse/discourse/blob/eb08c5453b00674ad39319a10c2fa461c607669c/app/assets/javascripts/discourse/app/controllers/discovery/topics.js#L119.

I explain this in the comment above.

Reply:
If the topic list is fetched with query params, and topic tracking state is synced to it, a hard refresh is needed to fix the topic tracking state.

Reply:
I have to say I don't really like it. Does this mean any time there are query parameters there will be two ajax requests on the server every time you view the topic list?

Reply:
@eviltrout No, this is only when `refresh` is called, after dismissing new topics when there are query params.

Reply:
Furthermore, all query params are cleared when `refresh` is called, except for params that are explicitly skipped (passed into the `refresh` function. 2 ajax calls will only ever happen when dismissing new and `?f=` or `?filter=` is present.

I agree with you it's smelly... But it would be so much more work messing with topic tracking state to avoid this.

Reply:
Oops I always click resolve conversation when I have a comment so it doesn't go through. This is cool. I wanted to mention refresh probably also happens if you click the logo, but that is not as common as navigating pages.